src/core/net/dns_client.cpp

```diff
@@ -257,9 +307,60 @@ Error Client::Response::FindServiceInfo(Section aSection, const Name &aName, Ser
 
 Error Client::AddressResponse::GetAddress(uint16_t aIndex, Ip6::Address &aAddress, uint32_t &aTtl) const
 {
-    return FindHostAddress(kAnswerSection, Name(*mQuery, kNameOffsetInQuery), aIndex, aAddress, aTtl);
+    Error error = kErrorNone;
+    Name  name(*mQuery, kNameOffsetInQuery);
+
+#if OPENTHREAD_CONFIG_DNS_CLIENT_NAT64_ENABLE
+
+    // If the response is for an IPv4 address query or if it is an IPv6
+    // address query response with an embedded IPv4 in its additional
+    // section, we read the IPv4 address and translate it to an IPv6
+    // address.
+
+    QueryInfo info;
+
+    info.ReadFrom(*mQuery);
+
+    if ((info.mQueryType == kIp4AddressQuery) || (mHasEmbeddedIp4))
+    {
+        Section     section;
+        ARecord     aRecord;
+        Ip6::Prefix nat64Prefix;
+
+        SuccessOrExit(error = GetNat64Prefix(nat64Prefix));
+
+        section = (info.mQueryType == kIp4AddressQuery) ? kAnswerSection : kAdditionalDataSection;
+        SuccessOrExit(error = FindARecord(section, name, aIndex, aRecord));
+
+        aAddress.SetFromTranslatedIp4Address(nat64Prefix, aRecord.GetAddress());
+        aTtl = aRecord.GetTtl();
+
+        ExitNow();
+    }
```

Comment:
In case this is an IPv6 address query and the response includes `AAAA` RR in the `Answer Section` and `A` RR in the `Additional Section`, current implementation with prefer the IPv4 address. Should we prefer IPv6 address for IPv6 query?

Reply:
I don't think that is the case in current code. If there is an `AAAA` record in the response, it will be preferred and used and NAT64 won't be used.

Note that the `if` block here is tied to `mHasEmbeddedIp4` being `true`. This boolean variable is only set when there is an error code from server (which indicates it could not provide `AAAA` record in response). Only in such a case, we check if there happens to be an `A` record in the received response and if so, we set `mHasEmbeddedIp4`.  This is all done in `ParseResponse()` method (new code in this PR).

Reply:
> This boolean variable is only set when there is an error code from server (which indicates it could not provide `AAAA` record in response).

I wonder what RCODE should the DNS server use when there is no AAAA RR available but A RR is available. This PR implies that RCODE should be NXDOMAIN. But I think it's also logical to use RCODE=NO_ERROR. I think it's better to search for A RR as long as there is no AAAA RR available, no matter what RCODE is.

Reply:
@abtink Thanks for clarification, I got the `mHasEmbeddedIp4` wrong! Maybe we can rename it to `mHasOnlyEmbeddedIp4` :)

Reply:
> But I think it's also logical to use RCODE=NO_ERROR. I think it's better to search for A RR as long as there is no AAAA RR available, no matter what RCODE is.

If I am not mistaken, I think if the server cannot provide any response for the main query question (which is `AAAA` records in this case) then it may not be allowed to use `NO_ERROR` response code.

We can add such a check in client to cover if server happens behave as this.

Let me see how complex it will be to add. If it does not add much overhead (code size, ect). I am ok adding it.

Reply:
I changed this in the new pushed commit, i.e. we don't check the response error code from server and instead check if the response does not contain an AAAA record for the query name.

Reply:
RCODE is defined to be NO_ERROR for a similar case:

https://tools.ietf.org/html/draft-vavrusa-dnsop-aaaa-for-free-00#section-5.3

But this draft also suggests to add to answer instead of additional section, which conflicts with this PR.

Maybe we should treat ANSWER and ADDITIONAL RRs as equal  so as to work with all server variations.

Reply:
Maybe @Abhayakara can chime in what RCODE is expected when no AAAA record is returned, but an A record is included in the ADDITIONAL section.

Reply:
If there is no AAAA record, but there are other answers on the name, the correct RCODE to return is "no error." The answer count of zero tells us that there's no data. The additional section is then available for synthesis.

Reply:
The draft is wrong, by the way. That would never get past the DNSOP working group's review.

Reply:
Perfect. Thanks guys!